lib/shared/bucketing-assembly-script/assembly/eventQueue/eventQueue.ts

```diff
@@ -0,0 +1,197 @@
+     * as payloads to be sent by the Native code again.
+     *
+     * If there are payloads that are still marked as `inProgress` when `flush()` is called, it should return an
+     * exception to the native code.
+     */
+    flush(): string {
+        const payloads: FlushPayload[] = this.requestPayloadManager.constructFlushPayloads(
+            this.userEventQueue,
+            this.aggEventQueue
+        )
+        this.userEventQueue = new Map<string, UserEventsBatchRecord>()
+        this.aggEventQueue = new Map<string, Map<string, Map<string, Map<string, i64>>>>()
+        return jsonArrFromValueArray(payloads).stringify()
+    }
+
+    onPayloadSuccess(payloadId: string): void {
+        this.requestPayloadManager.markPayloadSuccess(payloadId)
+    }
+
+    onPayloadFailure(payloadId: string, retryable: boolean): void {
+        this.requestPayloadManager.markPayloadFailure(payloadId, retryable)
+    }
+
+    queueEvent(user: DVCPopulatedUser, event: DVCEvent, featureVariationMap: Map<string, string>): void {
+        // TODO: Implement max queue size
+        // this.maxEventQueueSize = bucketedConfig?.project.settings.sdkSettings?.eventQueueLimit
+        // ?? this.maxEventQueueSize
+
+        const requestEvent = new DVCRequestEvent(event, user.user_id, featureVariationMap)
+        this.addEventToQueue(user, requestEvent)
+    }
+
+    private addEventToQueue(user: DVCPopulatedUser, event: DVCRequestEvent): void {
+        // TODO: implement?
+        // if (this.eventQueueSize() >= this.maxEventQueueSize) {
+        //     this.logger.warn(`Max event queue size reached, dropping event: ${event}`)
+        //     this.flushEvents()
+        //     return
+        // }
+
+        const user_id = user.user_id
+        let userEvents: UserEventsBatchRecord
+        if (!this.userEventQueue.has(user_id)) {
+            userEvents = new UserEventsBatchRecord(user, [])
+            this.userEventQueue.set(user_id, userEvents)
+        } else {
+            userEvents = this.userEventQueue.get(user_id)
+            userEvents.user = user
+        }
+        userEvents.events.push(event)
+    }
+
+    queueAggregateEvent(
+        event: DVCEvent,
+        variableVariationMap: Map<string, FeatureVariation>,
+        aggByVariation: boolean
+    ): void {
+        const type = event.type
+        const target = event.target
+        if (!target) {
+            throw new Error('Event missing target to save aggregate event')
+        }
+
+        // TODO: implement?
+        // if (this.eventQueueSize() >= this.maxEventQueueSize) {
+        //     this.logger.warn(`Max event queue size reached, dropping aggregate event: ${event}`)
+        //     this.flushEvents()
+        //     return
+        // }
+
+        let variableFeatureVarAggMap: Map<string, Map<string, Map<string, i64>>>
+        if (this.aggEventQueue.has(type)) {
```

Comment:
might be nice to break this method up into more manageable steps implemented as private functions, like:
`ensureHasType`, `ensureHasTarget`, `aggregateByVariation` etc